python/spark_sklearn/group_apply.py

```diff
@@ -0,0 +1,175 @@
+"""
+Method for applying arbitrary UDFs over grouped data.
+"""
+
+from distutils.version import LooseVersion
+from itertools import chain
+from py4j import java_gateway
+
+from pyspark.sql import DataFrame, SparkSession
+from pyspark.sql.functions import explode, udf
+from pyspark.sql.types import *
+from pyspark.sql.types import Row
+
+def gapply(grouped_data, func, schema, *cols):
+    """Applies the function `func` to data grouped by key. In particular, given a dataframe
+    grouped by some set of key columns key1, key2, ..., keyn, this method groups all the values
+    for each row with the same key columns into a single Pandas dataframe and by default invokes
+    `func((key1, key2, ..., keyn), values)` where the number and order of the key arguments is
+    determined by columns on which this instance's parent :class:`DataFrame` was grouped and
+    `values` is a `pandas.DataFrame` of columns selected by `cols`, in that order.
+
+    If there is only one key then the key tuple is automatically unpacked, with `func(key, values)`
+    called.
+
+    `func` is expected to return a `pandas.DataFrame` of the specified schema `schema`,
+    which should be of type :class:`StructType` (output columns are of this name and order).
+
+    If `spark.conf.get("spark.sql.retainGroupColumns")` is not `u'true'`, then `func` is
+    called with an empty key tuple.
+
+    If no `cols` are specified, then all grouped columns will be offered, in the order of the 
+    columns in the original dataframe. In either case, the Pandas columns will be named 
+    according to the DataFrame column names.
+
+    The order of the rows passed in as Pandas rows is not guaranteed to be stable relative to 
+    the original row order.
+
+    .. :note: Users must ensure that the grouped values for every group must fit entirely in
+    memory.
+    .. :note: This method is only available if Pandas is installed.
+
+    :param func: a two argument function, which may be either a lambda or named function
+    :param schema: the return schema for `func`, a :class:`StructType`
+    :param cols: list of column names (string only)
```

Comment:
do you mention that the other columns are dropped? It is obvious from the implementation, but it should be mentioned somewhere.